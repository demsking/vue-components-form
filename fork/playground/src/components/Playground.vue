<template>
  <div class="playground">
    <div class="playground__input">
      <div class="playground__input__schema">
        <h2>Schema</h2>
        <div class="playground__input__schema__editor">
          <PrismEditor
            :code="rawSchema"
            language="json"
            @change="setRawSchema"/>
        </div>
      </div>
      <div class="playground__input__rendering">
        <h2>Rendering</h2>
        <div ref="form" class="playground__input__rendering__viewport">
          <FormSchema :schema="schema" :descriptor="descriptor" v-model="model" @ready="generateCode" @submit.prevent>
            <div class="buttons">
              <button type="submit">Subscribe</button>
            </div>
          </FormSchema>
        </div>
      </div>
      <div class="playground__input__model">
        <h2>Model</h2>
        <pre>{{ model }}</pre>
      </div>
    </div>
    <div class="playground__output">
      <h2>Generated HTML</h2>
      <div class="playground__output__code">
        <PrismEditor :code="code" language="html" readonly/>
      </div>
    </div>
  </div>
</template>

<script>
  import '../libs/Prism';
  import '../styles/Prism.css';

  import { html } from 'js-beautify';

  import PrismEditor from 'vue-prism-editor';
  import FormSchema from '../../../dist/FormSchema.esm';
  import Schema from '../schema/newsletter';

  export default {
    components: { PrismEditor, FormSchema },
    data: () => ({
      model: undefined,
      rawSchema: JSON.stringify(Schema, null, 2),
      code: '',
      formatter: null,
      descriptor: {
        properties: {
          name: {
            properties: {
              first_name: {
                attrs: {
                  placeholder: 'Your First Name',
                  title: 'Please enter your first name'
                }
              },
              middle_name: {
                attrs: {
                  placeholder: 'Your Middle Name',
                  title: 'Please enter your middle name'
                }
              },
              last_name: {
                attrs: {
                  placeholder: 'Your Last Name',
                  title: 'Please enter your last name'
                }
              }
            }
          },
          email: {
            attrs: {
              placeholder: 'Your Email',
              title: 'Please enter your email'
            }
          },
          day: {
            label: 'Newsletter Day',
            attrs: {
              placeholder: 'Select your list subscription',
              title: 'Please select your list subscription'
            },
            labels: {
              monday: 'Monday',
              tuesday: 'Tuesday',
              wednesday: 'Wednesday',
              thursday: 'Thursday',
              friday: 'Friday',
              saturday: 'Saturday',
              sunday: 'Sunday'
            }
          },
          source: {
            kind: 'textarea',
            label: 'Source',
            description: 'Ex. Using the NPM Search Engine',
            attrs: {
              placeholder: 'How did you hear about us?'
            }
          },
          password: {
            attrs: {
              type: 'password'
            }
          },
          frequence: {
            labels: {
              daily: 'Daily',
              weekly: 'Weekly'
            }
          }
        }
      }
    }),
    computed: {
      schema() {
        return JSON.parse(this.rawSchema);
      }
    },
    methods: {
      setRawSchema(value) {
        try {
          JSON.parse(value);

          this.rawSchema = value;
        } catch(e) {}
      },
      generateCode() {
        const code = this.$refs.form.innerHTML;

        this.code = html(code.replace(/\s*<!---->/g, ''));
      }
    }
  }
</script>

<style>
  .playground {
    text-align: left;
    margin: auto;
    /* display: flex; */
    flex-direction: column;
    color: #f5f5f5;
    background-color: #333333;
    overflow: hidden;

    position: relative;
    height: 100%;
  }

  .playground h2 {
    margin: 0;
    padding: 10px;
    font-size: .90em;
    font-weight: 400;
    width: 100%;
  }

  .playground pre {
    margin: 0;
  }

  .playground__input {
    display: flex;
    flex-direction: row;
    /* max-height: 550px; */

    position: absolute;
    top: 0;
    left: 0;
    width: 100%;
    height: 60%;
  }

  .playground h2,
  .playground__input {
    border-bottom: 1px solid rgba(255, 255, 255, .1);
  }

  .playground__input__schema {
    display: flex;
    flex-direction: column;
  }

  .playground__input__schema__editor {
    flex: 1;
    display: flex;
    align-items: flex-start;
    max-width: 500px;
    overflow: auto;
  }

  .playground__input__rendering {
    border-left: 1px solid rgba(255, 255, 255, .1);
    border-right: 1px solid rgba(255, 255, 255, .1);
  }

  .playground__input__rendering,
  .playground__input__model {
    display: flex;
    flex-direction: column;
    align-items: flex-start;
    overflow: hidden;
  }

  .playground__input__rendering__viewport {
    padding: 20px;
    overflow: auto;
    margin: auto;
    color: #333;
    background-color: #f5f5f5;
  }

  .playground__input__rendering legend {
    font-size: 1.7em;
    text-align: center;
    margin-top: 0;
    margin-bottom: .2em
  }

  .playground__input__rendering legend + p {
    display: block;
    text-align: center;
    margin-bottom: 1.2em
  }

  .playground__input__rendering input,
  .playground__input__rendering textarea,
  .playground__input__rendering select {
    display: block;
  }

  .playground__input__model {
    margin: 0;
    min-width: 100px;
  }

  .playground__output {
    position: absolute;
    top: 60%;
    left: 0;
    width: 100%;
    height: 40%;

    overflow: hidden;
    display: flex;
    flex-direction: column;
  }

  .playground__output__code {
    flex: 1;
    display: flex;
    overflow: auto;
  }

  [data-fs-field] {
    display: flex;
    flex-direction: row;
    margin-bottom: 5px;
  }

  [data-fs-field] label {
    display: block;
    width: 120px;
    text-align: right;
    margin-right: 10px
  }

  [data-fs-kind="enum"] [data-fs-field] {
    flex-direction: row-reverse;
  }

  [data-fs-kind="enum"] [data-fs-field] label {
    text-align: left;
    flex: 1;
  }
</style>
