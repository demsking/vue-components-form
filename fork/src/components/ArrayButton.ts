import { FunctionalComponentOptions } from 'vue';

export const ArrayButton: FunctionalComponentOptions = {
  functional: true,
  name: 'ArrayButton',
  render(h, { data, listeners }) {
    return h('button', {
      attrs: { type: 'button', ...data.attrs },
      on: listeners
    }, 'Add');
  }
};
