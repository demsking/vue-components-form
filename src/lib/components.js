import { INPUT_TYPES } from './parser'

const BLOCK_TYPES = [INPUT_TYPES.TEXTAREA, INPUT_TYPES.SELECT]

const Input = {
  functional: true,
  render (h, { data, slots }) {
    const field = data.field
    const element = BLOCK_TYPES.includes(field.attrs.type)
      ? h(field.attrs.type, data, slots().default)
      : h('input', data)
    const nodes = [ element ]

    if (field.description) {
      nodes.push(h('small', field.description))
    }

    if (!field.label || (field.isArrayField && !BLOCK_TYPES.includes(field.attrs.type))) {
      if (nodes.length === 1) {
        return nodes[0]
      }
      return nodes
    }

    return h(Label, data, nodes)
  }
}

const CheckboxGroup = {
  functional: true,
  render (h, { data, slots }) {
    const name = data.field.attrs.name
    const description = data.field.description

    const children = [ ...slots().default ]

    if (description) {
      children.unshift(h('legend', description))

      delete data.field.description
    }

    return h('fieldset', { attrs: { name } }, children)
  }
}

const ArrayInputs = {
  functional: true,
  render (h, { slots }) {
    return h('div', {
      attrs: {
        'data-fs-array-inputs': true
      }
    }, slots().default)
  }
}

const Label = {
  functional: true,
  render (h, { data, slots }) {
    const field = data.field
    const attrs = {
      'data-fs-field': field.attrs.id,
      'data-fs-required': field.attrs.required
    }

    return h('div', { attrs }, [
      h('label', {
        attrs: {
          for: field.attrs.id
        }
      }, field.label),
      h('div', {
        attrs: {
          'data-fs-field-input': field.attrs.id
        }
      }, slots().default)
    ])
  }
}

const Fieldset = {
  functional: true,
  render (h, { data, slots }) {
    const field = data.field
    const newItemButton = data.newItemButton
    const nodes = []

    if (field.isArrayField) {
      if (newItemButton) {
        nodes.push(h(ArrayInputs, slots().default))
        nodes.push(h(ArrayButton, newItemButton))
      } else {
        nodes.push(h(CheckboxGroup, data, slots().default))
      }
    } else {
      nodes.push(h('fieldset', slots().default))
    }

    if (field.description) {
      nodes.push(h('small', field.description))
    }

    if (!field.label) {
      if (nodes.length === 1) {
        return nodes[0]
      }
      return nodes
    }

    return h(Label, data, nodes)
  }
}

const ArrayButton = {
  functional: true,
  render (h, { data }) {
    return h('button', {
      attrs: { type: 'button', ...data.props },
      on: data.on
    }, 'Add')
  }
}

const ErrorElement = {
  functional: true,
  render (h, { slots }) {
    return h('div', {
      attrs: {
        'data-fs-error': true
      }
    }, slots().default)
  }
}

const TAGS = {
  title: 'h1',
  description: 'p',
  formwrapper: 'div',
  form: 'form',
  text: Input,
  option: 'option',
  fieldset: Fieldset,
  error: ErrorElement
}

export class Components {
  constructor () {
    this.$ = {}

    for (let type in TAGS) {
      this.set(type, TAGS[type], true)
    }
  }

  set (type, component, native = false) {
    this.$[type] = { type, native, component }
  }

  input ({ field, fieldParent = null, listeners = {} }) {
    const { type } = field.attrs
    const element = field.hasOwnProperty('items') && groupedArrayTypes.includes(type)
      ? this.$.fieldset
      : this.$[type] || this.$[INPUT_TYPES.TEXT]

    const data = {
      field,
      fieldParent,
      components: this.components,
      attrs: {},
      props: {},
      domProps: {},
      [argName(element)]: { ...field.attrs },
      on: listeners
    }

    return { element, data }
  }
}

export function argName (el) {
  return el.native ? 'attrs' : 'props'
}

export const groupedArrayTypes = [
  INPUT_TYPES.RADIO, INPUT_TYPES.CHECKBOX, 'input', INPUT_TYPES.TEXTAREA
]

export const fieldTypesAsNotArray = [
  INPUT_TYPES.RADIO, INPUT_TYPES.TEXTAREA, INPUT_TYPES.SELECT
]

export const inputName = (field, index) => `${field.attrs.name}-${index}`
