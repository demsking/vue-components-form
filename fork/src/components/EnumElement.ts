import { FunctionalComponentOptions } from 'vue';
import { HelperElement } from "@/components/HelperElement";

export const EnumElement: FunctionalComponentOptions = {
  name: 'EnumElement',
  functional: true,
  render(h, { data, props }) {
    const { components } = props;

    const nodes = props.field.fields.map(({ field }) => {
      const type = field.kind;

      return h(components.get(type), {
        attrs: field.attrs.input,
        props: {
          field: field,
          value: field.model,
          components: props.components
        }
      });
    });

    const helper = h(HelperElement, data);

    if (helper) {
      nodes.unshift(helper);
    }

    if (props.field.label) {
      const legend = h('legend', {
        attrs: props.field.attrs.label
      }, props.field.label);

      nodes.unshift(legend);
    }

    return h('fieldset', {
      attrs: props.field.attrs.input
    }, nodes);
  }
};
