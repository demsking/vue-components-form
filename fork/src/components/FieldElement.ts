import { HelperElement } from "@/components/HelperElement";
import { Dictionary, FieldComponent } from '@/types';

export const FieldElement: FieldComponent = {
  name: 'FieldElement',
  functional: true,
  render(h, { data, props, slots }) {
    const attrs: Dictionary = {
      'data-fs-kind': props.field.kind,
      'data-fs-field': props.field.name,
      'data-fs-required': props.field.required
    };

    const labelElement = h('label', {
      attrs: props.field.attrs.label
    }, props.field.descriptor.label);

    const fieldNodes = [
      ...slots().default,
      h(HelperElement, data)
    ];

    const fieldElement = h('div', {
      attrs: {
        'data-fs-input': (props.field.attrs.input as any).type || props.field.kind
      }
    }, fieldNodes);

    const nodes = [ labelElement, fieldElement ];

    return h('div', { attrs }, nodes);
  }
};
