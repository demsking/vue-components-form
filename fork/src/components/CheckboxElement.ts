import { CreateStateInput } from '@/lib/NativeElementsLib';

export const CheckboxElement = CreateStateInput('checkbox');
