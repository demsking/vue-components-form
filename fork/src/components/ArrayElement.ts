import { FunctionalComponentOptions } from 'vue';
import { FieldElement } from './FieldElement';
import { ArrayButton } from './ArrayButton';

function renderField(h, field, props) {
  return h(field.component, {
    attrs: field.attrs.input,
    props: {
      ...props,
      field: field,
      value: field.model
    }
  });
}

export const ArrayElement: FunctionalComponentOptions = {
  name: 'ArrayElement',
  functional: true,
  render(h, { props }) {
    const attrs = props.field.attrs.input;
    const nodesx = props.field.items.map((field) => {
      return h(field.component, {
        attrs: field.attrs.input,
        props: {
          ...props,
          field: field,
          value: field.model
        }
      });
    });

    const max = props.field.maxItems
      ? props.field.maxItems + props.field.additionalItems.length
      : -1;

    const limit = max === -1
      ? props.field.count
      : props.field.count < props.field.maxItems
        ? props.field.count
        : props.field.maxItems;

    const nodes = Array(...Array(limit)).map((x, i) => {
      const field = props.field.items[0];

      return renderField(h, field, props);
    });

    if (limit < props.field.count) {
      props.field.additionalItems.forEach((field) =>  {
        nodes.push(renderField(h, field, props));
      });
    }

    const addButtonElement = h(ArrayButton, {
      attrs: {
        disabled: props.field.count === max
      },
      props: props,
      on: {
        click() {
          props.field.count++;
        }
      }
    });

    nodes.push(addButtonElement);

    return h(FieldElement, { attrs, props }, nodes);
  }
};
