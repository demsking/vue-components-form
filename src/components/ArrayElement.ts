import { VNode } from 'vue';
import { ArrayComponent } from '@/types';
import { ArrayButton } from '@/components/ArrayButton';
import { FieldElement } from '@/components/FieldElement';
import { FieldsetElement } from '@/components/FieldsetElement';

export const ArrayElement: ArrayComponent = {
  name: 'ArrayElement',
  functional: true,
  render(h, { data, props }) {
    const nodes: any = props.descriptor.children
      .map((descriptor) => {
        const field = descriptor.field;
        const buttonsWrapper: VNode[] = [];

        if (!props.field.uniqueItems && props.field.sortable) {
          const buttonsNodes = descriptor.buttons.map((button) => h(ArrayButton, {
            props: button
          }));

          buttonsWrapper.push(h('div', {
            key: props.field.key + field.key,
            attrs: {
              'data-fs-buttons': buttonsNodes.length
            }
          }, buttonsNodes));
        }

        if (descriptor.kind === 'object') {
          return h(FieldElement, {
            props: {
              descriptor,
              field: { ...field, helper: {} } // remove the duplicate bottom helper
            }
          }, [
            h(descriptor.component, {
              props: { field, descriptor }
            }),
            buttonsWrapper
          ]);
        }

        return h(descriptor.component, {
          attrs: descriptor.attrs,
          props: { field, descriptor }
        }, buttonsWrapper);
      });

    if (!props.field.uniqueItems) {
      nodes.push(h(ArrayButton, {
        props: props.descriptor.pushButton
      }));
    }

    return h(FieldsetElement, data, nodes);
  }
};
