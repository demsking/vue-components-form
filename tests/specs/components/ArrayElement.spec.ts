import { mount } from '@vue/test-utils';
import { ArrayElement } from '@/components/ArrayElement';
import { Options } from '../../lib/Options';

describe('components/ArrayElement', () => {
  it('should successfully render component with an empty array', () => {
    const { context } = Options.get({
      id: 'id-characters',
      name: 'characters',
      schema: {
        type: 'array',
        title: 'Empty array',
        description: 'Your characters'
      }
    });

    const wrapper = mount(ArrayElement, { context });
    const expected = '<fieldset id="id-characters" name="characters" aria-labelledby="id-characters-label" aria-describedby="id-characters-helper"><legend id="id-characters-label" for="id-characters">Empty array</legend><p id="id-characters-helper">Your characters</p><button type="button" disabled="disabled" data-fs-button="push">+</button></fieldset>';

    expect(wrapper.html()).toBe(expected);
  });

  it('should successfully render component with an empty array with a defined model', () => {
    const { context } = Options.get({
      id: 'id-characters',
      name: 'characters',
      schema: {
        type: 'array',
        title: 'Empty array',
        description: 'Your characters'
      },
      model: ['freezer']
    });

    const wrapper = mount(ArrayElement, { context });
    const expected = '<fieldset id="id-characters" name="characters" aria-labelledby="id-characters-label" aria-describedby="id-characters-helper"><legend id="id-characters-label" for="id-characters">Empty array</legend><p id="id-characters-helper">Your characters</p><button type="button" disabled="disabled" data-fs-button="push">+</button></fieldset>';

    expect(wrapper.html()).toBe(expected);
  });

  it('should successfully render component with an array schema with an empty model', () => {
    const { context } = Options.get({
      id: 'id-characters',
      name: 'characters',
      schema: {
        type: 'array',
        title: 'Characters',
        description: 'Your characters',
        minItems: 1,
        maxItems: 3,
        items: { type: 'string' }
      },
      model: []
    });

    const wrapper = mount(ArrayElement, { context });
    const expected = '<fieldset id="id-characters" name="characters" aria-labelledby="id-characters-label" aria-describedby="id-characters-helper"><legend id="id-characters-label" for="id-characters">Characters</legend><p id="id-characters-helper">Your characters</p><div data-fs-kind="string" data-fs-type="text" data-fs-field="characters"><label for="id-characters-0"></label><div data-fs-input="text"><input id="id-characters-0" type="text" name="characters"><div data-fs-buttons="3"><button type="button" disabled="disabled">↑</button><button type="button" disabled="disabled">↓</button><button type="button">-</button></div></div></div><button type="button" data-fs-button="push">+</button></fieldset>';

    expect(wrapper.html()).toMatchSnapshot(expected);
  });

  it('should successfully render component with an array schema with a defined model', () => {
    const { context } = Options.get({
      id: 'id-characters',
      name: 'characters',
      schema: {
        type: 'array',
        title: 'Characters',
        description: 'Your characters',
        minItems: 1,
        maxItems: 3,
        items: { type: 'string' }
      },
      model: ['Goku', 'Freezer']
    });

    const wrapper = mount(ArrayElement, { context });
    const expected = '<fieldset id="id-characters" name="characters" aria-labelledby="id-characters-label" aria-describedby="id-characters-helper"><legend id="id-characters-label" for="id-characters">Characters</legend><p id="id-characters-helper">Your characters</p><div data-fs-kind="string" data-fs-type="text" data-fs-field="characters"><label for="id-characters-0"></label><div data-fs-input="text"><input id="id-characters-0" type="text" name="characters" value="Goku"><div data-fs-buttons="3"><button type="button" disabled="disabled">↑</button><button type="button">↓</button><button type="button">-</button></div></div></div><div data-fs-kind="string" data-fs-type="text" data-fs-field="characters"><label for="id-characters-1"></label><div data-fs-input="text"><input id="id-characters-1" type="text" name="characters" value="Freezer"><div data-fs-buttons="3"><button type="button">↑</button><button type="button" disabled="disabled">↓</button><button type="button">-</button></div></div></div><button type="button" data-fs-button="push">+</button></fieldset>';

    expect(wrapper.html()).toMatchSnapshot(expected);
  });

  it('clicking to the push button should successfully trigger the options.onChange callback', () => {
    const { context, parser } = Options.get({
      id: 'id-characters',
      name: 'characters',
      schema: {
        type: 'array',
        title: 'Characters',
        description: 'Your characters',
        minItems: 1,
        maxItems: 3,
        items: { type: 'string' }
      },
      model: ['Goku', 'Freezer']
    });

    const wrapper = mount(ArrayElement, { context });
    const button = wrapper.find('button[data-fs-button=push]');

    expect(parser.rawValue).toEqual(['Goku', 'Freezer']);

    button.trigger('click');

    expect(parser.rawValue).toEqual(['Goku', 'Freezer', undefined]);
  });

  describe('should successfully render enum array schema', () => {
    const onChangeMock = jest.fn();
    const { context, schema } = Options.get({
      id: 'id-characters',
      name: 'characters',
      schema: {
        type: 'array',
        title: 'Characters',
        description: 'Your favorite characters',
        items: {
          type: 'string',
          enum: ['Goku', 'Gohan', 'Vegeta', 'Picolo']
        },
        uniqueItems: true
      },
      model: ['Goku', 'Picolo'],
      onChange: onChangeMock
    });

    const wrapper = mount(ArrayElement, { context });
    const pushButton = wrapper.find('button[data-fs-button=push]');
    const inputs: any = wrapper.findAll('input[type=checkbox]');

    it('should don\'t have push button', () => {
      expect(pushButton.exists()).toBeFalsy();
    });

    it('should have exaltly schema.items.enum.length input[type=checkbox]', () => {
      expect(inputs.length).toBe(schema.items.enum.length);
    });

    it('should have right initial model', () => {
      const [ [ initialValue ] ] = onChangeMock.mock.calls;

      expect(initialValue).toEqual(['Goku', 'Picolo']);
      expect(onChangeMock.mock.calls.length).toBe(1);
    });

    it('should have updated model after interaction', () => {
      inputs.at(1).setChecked(true);

      const [ [], [ changedValue ] ] = onChangeMock.mock.calls;

      expect(changedValue).toEqual(['Goku', 'Gohan', 'Picolo']);
      expect(onChangeMock.mock.calls.length).toBe(2);
    });
  });
});
