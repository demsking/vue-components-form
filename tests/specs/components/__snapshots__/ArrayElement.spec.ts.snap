// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/ArrayElement should successfully render component with an array schema with a defined model: <fieldset id="id-characters" name="characters" aria-labelledby="id-characters-label" aria-describedby="id-characters-helper"><legend id="id-characters-label" for="id-characters">Characters</legend><p id="id-characters-helper">Your characters</p><div data-fs-kind="string" data-fs-type="text" data-fs-field="characters"><label for="id-characters-0"></label><div data-fs-input="text"><input id="id-characters-0" type="text" name="characters" value="Goku"><div data-fs-buttons="3"><button type="button" disabled="disabled">↑</button><button type="button">↓</button><button type="button">-</button></div></div></div><div data-fs-kind="string" data-fs-type="text" data-fs-field="characters"><label for="id-characters-1"></label><div data-fs-input="text"><input id="id-characters-1" type="text" name="characters" value="Freezer"><div data-fs-buttons="3"><button type="button">↑</button><button type="button" disabled="disabled">↓</button><button type="button">-</button></div></div></div><button type="button" data-fs-button="push">+</button></fieldset> 1`] = `
<fieldset id="id-characters" name="characters" aria-labelledby="id-characters-label" aria-describedby="id-characters-helper">
  <legend id="id-characters-label" for="id-characters">Characters</legend>
  <p id="id-characters-helper">Your characters</p>
  <div data-fs-kind="string" data-fs-type="text" data-fs-field="characters"><label for="id-characters-0"></label>
    <div data-fs-input="text"><input id="id-characters-0" type="text" name="characters" value="Goku">
      <div data-fs-buttons="3"><button type="button" disabled="disabled">↑</button><button type="button">↓</button><button type="button">-</button></div>
    </div>
  </div>
  <div data-fs-kind="string" data-fs-type="text" data-fs-field="characters"><label for="id-characters-1"></label>
    <div data-fs-input="text"><input id="id-characters-1" type="text" name="characters" value="Freezer">
      <div data-fs-buttons="3"><button type="button">↑</button><button type="button" disabled="disabled">↓</button><button type="button">-</button></div>
    </div>
  </div><button type="button" data-fs-button="push">+</button>
</fieldset>
`;

exports[`components/ArrayElement should successfully render component with an array schema with an empty model: <fieldset id="id-characters" name="characters" aria-labelledby="id-characters-label" aria-describedby="id-characters-helper"><legend id="id-characters-label" for="id-characters">Characters</legend><p id="id-characters-helper">Your characters</p><div data-fs-kind="string" data-fs-type="text" data-fs-field="characters"><label for="id-characters-0"></label><div data-fs-input="text"><input id="id-characters-0" type="text" name="characters"><div data-fs-buttons="3"><button type="button" disabled="disabled">↑</button><button type="button" disabled="disabled">↓</button><button type="button">-</button></div></div></div><button type="button" data-fs-button="push">+</button></fieldset> 1`] = `
<fieldset id="id-characters" name="characters" aria-labelledby="id-characters-label" aria-describedby="id-characters-helper">
  <legend id="id-characters-label" for="id-characters">Characters</legend>
  <p id="id-characters-helper">Your characters</p>
  <div data-fs-kind="string" data-fs-type="text" data-fs-field="characters"><label for="id-characters-0"></label>
    <div data-fs-input="text"><input id="id-characters-0" type="text" name="characters">
      <div data-fs-buttons="3"><button type="button" disabled="disabled">↑</button><button type="button" disabled="disabled">↓</button><button type="button">-</button></div>
    </div>
  </div><button type="button" data-fs-button="push">+</button>
</fieldset>
`;
